zscript/Weapons/ThrownGrenades.zs

```diff
@@ -0,0 +1,155 @@
+class PB_ThrownGrenade : Actor
+{
+    int grenadeTicks, grenadeTime, grenadeBeepSpeed, exploRandom;
+
+    private int zscriptTick;
+    flagdef DOZSCRIPTTICK: zscriptTick, 0;
+
+    Default {
+        Radius 2;
+        Height 3;
+        Projectile;
+        Speed 30;
+        Damage 0;
+        Gravity 0.7;
+        Scale 0.2;
+        +MISSILE;
+        -NOGRAVITY;
+        -BLOODSPLATTER;
+        +EXPLODEONWATER;
+        +SKYEXPLODE;
+        +DOOMBOUNCE;
+        +FLOORCLIP;
+        +DONTGIB;
+        -NOBLOCKMAP;
+        Species "Marine";
+        +MTHRUSPECIES;
+        +THRUSPECIES;
+        +DontHarmSpecies;
+        +UseBounceState;
+        +RollSprite;
+        +RollCenter;
+        DamageType "ExplosiveImpact";
+        BounceFactor 0.5;
+        WallBounceFactor 0.25;
+        Health 5;
+        SeeSound "GBOUNCE";
+        DeathSound "none";
+        Obituary "%o gobbled %k's grenade.";
+        +PB_ThrownGrenade.DOZSCRIPTTICK;
+    }
+
+	override void Tick()
+    {
+        Super.Tick();
+
+        //console.printf("%i %i %i %i", grenadeTicks, grenadeBeepSpeed, exploRandom, grenadeTime);
+
+        if(bDOZSCRIPTTICK)
+        {
+            GrenadeTicks++;
+
+            if(GrenadeTicks % 8 == 0)                
+                GrenadeTime++;
+
+            if(GrenadeTicks % 3 == 0 && GrenadeBeepSpeed > 1)
+                grenadeBeepSpeed--;
+
+            if(GrenadeTicks % 4 == 0)

```

Comment:
I am genuinely confused. I'm sorry, I promise I don't want to sound rude or something. And maybe this is not the place to talk about this kind of things because you are just modding an engine with the tools tou have available. But, as a new commer to the doom modding scene and as a senior software developer, this looks really ugly to me. Is this the standarized `ZScript` way of dealing with stuff like this? Or is there a better way?

If this is the only way, then you guys deserve a medal.

I mean, every class, has to implement it's own `Tick` method and has to manually call `Super.Tick()`, and keep track of the "Ticks" or "Time" that has transcurred? Isn't there any async/timer/scheduler API to delay the execution of certain functions? Or a Future/Promises API? Something like `setTimeout` or even a `sleep` function or something to get notified once certain time/ticks has passed? The "cron" scheduler has been with us since 1974 if I'm not wrong. A simple "tick" event bus with the option that your callback gets executed once certain amounts of ticks have transcurred would be enough. There are simply no excuses that you have to work with these tools. in 2023

Also, the whole idea of depending on abstract game logical ticks frightens me. Decoupling the most game logic as possible from the main loop and from a fixed tick rate is something that videogames have been doing or trying to do for like 25 years. Time-based solutions, event-driven logic, etc are a more straighforward solution and the industry standard. As far as I know, from peeking into GZDoom source code for about three days, is that the game *tries* to run at ¿36? Hz ticks rate But there is no guarantee at all that that the "tick" function is going to be executed 36 times a second. Nor that it does get called regularly with the same interval. In fact, it doesn't, probably even if our PC runs the game at higher than 36 FPS. Computationally it is impossible that to guarantee that, considering GZDoom seems to be single threaded, and so far I have not seen any use of fibers, coroutines, or  any other asynchronous non-blocking mechanisms.

I'm sorry for bringing this stuff here, this is probably not the place for this discussions. It's just that the deeper I dive, the uglier creatures I see. Anyway, despite the GZDoom/ZScript ugliness, the PR looks fine to me.

Reply:
no